src/Unix-Timestamp/time2unixts.1
.\" ==========================================================================
.\"         ____            _                     _____           _
.\"        / ___| _   _ ___| |_ ___ _ __ ___     |_   _|__   ___ | |___
.\"        \___ \| | | / __| __/ _ \ '_ ` _ \ _____| |/ _ \ / _ \| / __|
.\"         ___) | |_| \__ \ ||  __/ | | | | |_____| | (_) | (_) | \__ \
.\"        |____/ \__, |___/\__\___|_| |_| |_|     |_|\___/ \___/|_|___/
.\"               |___/
.\"                             --- System-Tools ---
.\"                  https://www.nntb.no/~dreibh/system-tools/
.\" ==========================================================================
.\"
.\" Time to Unix Timestamp Conversion
.\" Copyright (C) 2026 by Thomas Dreibholz
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" Contact: thomas.dreibholz@gmail.com
.\"
.\" ###### Setup ############################################################
.Dd September 14, 2026
.Dt time2unixts 1
.Os time2unixts
.\" ###### Name #############################################################
.Sh NAME
.Nm time2unixts
.Nd converts date/time strings to Unix timestamps
.\" ###### Synopsis #########################################################
.\" Manpage syntax help:
.\" https://forums.freebsd.org/threads/howto-create-a-manpage-from-scratch.13200/
.Sh SYNOPSIS
.Nm time2unixts
.Op Fl T Ar time\_format\_template | Fl Fl template Ar time\_format\_template
.br
.Op Fl F | Fl Fl float
.Op Fl I | Fl Fl integer\-decimal
.Op Fl X | Fl Fl integer\-hexadecimal
.Op Fl 0 | Fl Fl integer\-0x\-hexadecimal
.br
.Op Fl H | Fl Fl human\-readable
.br
.Op Fl s | Fl Fl seconds
.Op Fl m | Fl Fl milliseconds
.Op Fl u | Fl Fl microseconds
.Op Fl n | Fl Fl nanoseconds
.br
time\_string
.Op ...
.Nm time2unixts
.Op Fl h | Fl Fl help
.Nm time2unixts
.Op Fl v | Fl Fl version
.\" ###### Description ######################################################
.Sh DESCRIPTION
.Nm time2unixts
converts date/time strings in the format %Y\-%m\-%d %H:%M:%S in UTC to Unix timestamps, which is the time (nanoseconds, microseconds, milliseconds, or seconds) since the Unix Epoch, i.e. January 1, 1970, 00:00:00.000000000 UTC.
.Pp
.\" ###### Arguments ########################################################
.Sh ARGUMENTS
The following options are available:
.Bl -tag -width indent
.It time\_string ...
The date/time strings to be converted, as %Y\-%m\-%d %H:%M:%S[.SSSSSSSSS]. The seconds part may be fractional; the fraction is optional. If no date/time string is provided, the current system time is used.
.It Fl T Ar time\_format\_template | Fl Fl template Ar time\_format\_template
Sets the template for parsing the date/time strings. Default is \(dq"%Y-%m-%d %H:%M:%S\(dq. The formatting is according to strptime(). See
.Xr strptime 3
for details.
.It Fl F | Fl Fl float
Print the Unix timestamp as a floating-point number.
.It Fl I | Fl Fl integer\-decimal
Print the Unix timestamp as a decimal integer number. This is the default.
.It Fl X | Fl Fl integer\-hexadecimal
Print the Unix timestamp as a hexadecimal integer number.
.It Fl 0 | Fl Fl integer\-0x\-hexadecimal
Print the Unix timestamp as a hexadecimal integer number, prefixed by 0x.
.It Fl H | Fl Fl human\-readable
Provide the output in human\-readable form. Default: off.
.It Fl s | Fl Fl seconds
The Unix timestamp is printed in seconds. Default is in nanoseconds.
.It Fl m | Fl Fl milliseconds
The Unix timestamp is printed in milliseconds. Default is in nanoseconds.
.It Fl u | Fl Fl microseconds
The Unix timestamp is printed in microseconds. Default is in nanoseconds.
.It Fl n | Fl Fl nanoseconds
The Unix timestamp is printed in nanoseconds. This is the default.
.It Fl h | Fl Fl help
Prints the program's command\-line parameters and exits.
.It Fl v | Fl Fl version
Prints the program version and exits.
.El
.\" ###### Exit status ######################################################
.Sh EXIT STATUS
The
.Nm
tool exits with 0 on success, and >0 in case of an error.
.\" ###### Examples #########################################################
.Sh EXAMPLES
.Bl -tag -width indent
.It time2unixts
.It time2unixts \-\-integer\-0x\-hexadecimal
.It time2unixts \-\-integer\-hexadecimal \-\-seconds
.It time2unixts \(dq2026\-03\-17 10:52:35\(dq \-\-seconds
.It time2unixts \(dq1969\-12\-20 10:13:20\(dq
.It time2unixts \(dq1955\-12\-24 22:22:22.000\(dq \-\-milliseconds
.It time2unixts \(dq2026\-03\-17 08:31:19\(dq \-\-human\-readable \-\-seconds
.It time2unixts \(dq2026\-03\-17 08:31:19\(dq \-s
.It time2unixts \(dq1976\-09\-29 12:12:03.123\(dq \-m
.It time2unixts \-s \(dq2026-09-04 19:30:14.7787085\(dq
.It time2unixts \-T \(dq%b %d %H:%M:%S %Y %Z\(dq \(dqJun  4 19:01:25 2036 GMT\(dq
.It time2unixts \-T \(dq%b %d %H:%M:%S %Y %Z\(dq \(dqSep 29 13:10:12.567890123 1976 CET\(dq
.It time2unixts \-\-version
.It time2unixts \-\-help
.El
.\" ###### See also #########################################################
.Sh SEE ALSO
.Xr strptime 3
.Xr unixts2time 1
.\" ###### Notes ############################################################
.Sh NOTES
This program is part of System\-Tools. The latest version of System\-Tools can be found on the System\-Tools Homepage at
.Lk https://www.nntb.no/\(tidreibh/system\-tools/ "System\-Tools Homepage" .
.\" ###### Authors ##########################################################
.Sh AUTHORS
Thomas Dreibholz,
.Lk https://www.nntb.no/\(tidreibh "Homepage"


ChangeLog
commit 0273e844adfe18591857f25a700da1efafae3566
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 17:05:08 2026 +0200

    Updated PO files.

commit 40d566574253e104ef1d8262409528e73e13f313
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 17:03:38 2026 +0200

    Bugfix: system-info.d and system-maintenance.d under Homebrew contain symlinks, not files.
    
    Also improved handling of script directories.

commit 08003be7000f240929b940964708ab664964dc6f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 16:08:34 2026 +0200

    Updated FreeBSD port.

commit 8afc7ab571b6d051f303bb02bdafd55c5e08a03c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 8 12:25:01 2026 +0200

    Updated APKBUILD file.

commit 5d2eb9cc96252ebc46630d39f9e3fd6c4dff49ab
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 18:22:22 2026 +0200

    Updated APKBUILD file.

commit 587ce77acef704b83f6d1452451b0100f036da9b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 18:13:14 2026 +0200

    New release td-system-tools-2.7.13.

commit 8d79add36e806c3976b3860e870e3bec61681aae
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Sep 4 18:09:04 2026 +0200

    PO file updates.

commit 700ba8a7c02336395182b8f2004a206dd990e29e
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Wed Sep 2 21:28:51 2026 +0200

    Updated FreeBSD port.

commit ab19df84aaf7d7e7a7f8a90faad8e7162ef0de11
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Sep 1 16:23:43 2026 +0200

    Updated APKBUILD file.

commit 46ec3f27549ea59e0a5ee86e3ded61ca8a367eb7
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Aug 31 16:27:43 2026 +0200

    New release td-system-tools-2.7.12.

commit 0b219ca51062396c85d114933be2e6d0361fe1f8
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Aug 27 09:28:47 2026 +0000

    Improved SSHFP RR output.

commit c3c0f692061daee599598b41d1244e70f8890d06
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Thu Aug 27 09:00:11 2026 +0000

    Bugfix: Fixed querying the swap size under Solaris.

commit 54d2dbba2c6d4f1ba16b8ddf21a76356750f719f
Merge: 2e48c7d f7d36f8
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Aug 11 14:52:15 2026 +0000

    WIP on master: 2e48c7d Updated FreeBSD port.

commit 2e48c7d12ac602575a36ce4d19e53f700de8b92f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 13:49:50 2026 +0200

    Updated FreeBSD port.

commit 2d15c6435117cd7344777bb282d09f30481b08da
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 13:21:42 2026 +0200

    Updated APKBUILD file.

commit 7a87b24841deddc5c6d5c27f9718001667375054
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 12:49:30 2026 +0200

    New release td-system-tools-2.7.11.

commit 38752f318e2dec60f918678400c8cb7755e042eb
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 11:26:42 2026 +0200

    Further clean-ups.

commit 776e793eca014c49a661105c628b399fd203b235
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 11:24:23 2026 +0200

    Further ShellCheck clean-ups.

commit 660a4174809a25da69ad67ca390466e376cc1d53
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 11:19:56 2026 +0200

    Minor clean-up.

commit 96983e49205334ba499846d12cff8a7675d1ca92
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 31 11:19:06 2026 +0200

    Improved compatibility of ShellCheck script.

commit 4b779c0d543f127805a5cc7f95e7ffd21d6f78f9
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 15:11:18 2026 +0200

    Updated PO files.

commit ef723836537074580f41524970d0c76358e18893
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 13:08:31 2026 +0000

    Solaris UTF-8 character width computation fix.

commit 5ba247e608aae3bbb0badde0866172f113a82da4
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 14:34:46 2026 +0200

    Updated translations.

commit 23f14ac5698a53e8a330412b7a4a52739c259c7c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 14:28:12 2026 +0200

    Improved checks for Bash version and enhanced getopt.

commit 0e00b3fb1e1323fdeaac952cd9a5c522c1c109df
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 12:58:39 2026 +0200

    Also need to set TEXTDOMAINDIR in scripts explicitly.

commit 39a99cfe9e7669d92e62245a8a66629a64b8e567
Merge: 727a9c6 d13052b
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 12:42:43 2026 +0200

    Merge branch 'master' of github.com:dreibh/system-tools

commit 727a9c6306d6073b41242b26511e344754a0a62f
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Tue Jul 28 12:41:57 2026 +0200

    Explicitly set locale directory in bindtextdomain() call.

commit d13052bb7adb6d57d5fc94a246932d0da8100f9c
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Jul 27 20:49:26 2026 +0000

    Updated comment.

commit 354674a0425a25975e2ccfb6dfa8e408dd705c98
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Jul 27 20:46:58 2026 +0000

    Added Solaris battery status block.

commit 7394954f5f0000425372b73f736e9711fede8b04
Merge: 8d67bd5 825fcf2
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Jul 27 15:45:14 2026 +0200

    Merge branch 'master' of github.com:dreibh/system-tools

commit 8d67bd5f749c54f04b5f2dddf6f2d8799f7a7565
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Jul 27 15:45:00 2026 +0200

    Improved example.

commit c656075f9f9d427c81b2ef66648436c53ebd3361
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Mon Jul 27 15:44:22 2026 +0200

    Added test script.

commit 825fcf2551699165b552a0d873f8e1e797ff2d8a
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 24 21:21:11 2026 +0000

    Fixed Debian/armhf build issue.

commit 3acd852be15d3d35f2e9971981e14ea106418e14
Author: Thomas Dreibholz <dreibh@simula.no>
Date:   Fri Jul 24 22:36:45 2026 +0200

    CMakeLists.txt update to use C23.

commit 99d67b708afe8e44a8edbf119aae5e2b46045aae

src/Unix-Timestamp/time2unixts.c
// ==========================================================================
//         ____            _                     _____           _
//        / ___| _   _ ___| |_ ___ _ __ ___     |_   _|__   ___ | |___
//        \___ \| | | / __| __/ _ \ '_ ` _ \ _____| |/ _ \ / _ \| / __|
//         ___) | |_| \__ \ ||  __/ | | | | |_____| | (_) | (_) | \__ \.
//        |____/ \__, |___/\__\___|_| |_| |_|     |_|\___/ \___/|_|___/
//               |___/
//                             --- System-Tools ---
//                  https://www.nntb.no/~dreibh/system-tools/
// ==========================================================================
//
// Time to Unix Timestamp Conversion
// Copyright (C) 2026 by Thomas Dreibholz
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// Contact: thomas.dreibholz@gmail.com

#define _GNU_SOURCE
#define __EXTENSIONS__
#include <ctype.h>
#include <errno.h>
#include <getopt.h>
#include <locale.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#ifdef ENABLE_NLS
#include <libintl.h>
#else
#define bindtextdomain(domain, dirname) { }
#define textdomain(domain) { }
#define gettext(string) string
#define ngettext(singular, plural, n) ((n) == 1 ? (singular) : (plural))
#endif

#include "package-version.h"

#if defined(__STDC_VERSION__) && (__STDC_VERSION__ < 202311L)
#ifndef nullptr
#define nullptr ((void*)0)
#endif
#endif


// ###### Find %S placeholder for seconds in time format string #############
static const char* findSecondsPlaceholder(const char* formatString)
{
   const char* ptr = formatString;
   while(*ptr) {
      if(*ptr == '%') {
         switch(*(ptr + 1)) {
            case 0x00:
               ptr++;
             break;
            case 'S':
               return ptr;
             break;
            default:
               ptr += 2;
             break;
         }
      } else {
         ptr++;
      }
   }
   return nullptr;
}


// ###### Print integer value ###############################################
static void printBigInteger(
#if (defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 202311L) && defined(__BITINT_MAXWIDTH__) && (__BITINT_MAXWIDTH__ >= 128))
   _BitInt(128) value,
#else
   long long    value,
#endif
   const int    base,
   const bool   withPrefix)
{
   if(value < 0) {
      putchar('-');
   }
   if( (withPrefix) && (base == 16)) {
      fputs("0x", stdout);
   }

   char buffer[64];
   int  idx = 0;
   if(value < 0) {   // Negative value
      do {
         int remainder = (int)(value % base);
         if(remainder < 0) remainder = -remainder;
         buffer[idx++] = (remainder < 10) ? (remainder + '0') : (remainder - 10 + 'a');
         value /= base;
      } while (value < 0);
   }
   else {            // Non-negative value
      do {
         int remainder = (int)(value % base);
         buffer[idx++] = (remainder < 10) ? (remainder + '0') : (remainder - 10 + 'a');
         value /= base;
      } while (value > 0);
   }

   while (idx > 0) {
      putchar(buffer[--idx]);
   }
}


// ###### Version ###########################################################
#if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 202000L)
[[ noreturn ]]
#endif
static void version(void)
{
   printf("time2unixts %s\n", SYSTEMTOOLS_VERSION);
   exit(0);
}


// ###### Usage #############################################################
#if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 202000L)
[[ noreturn ]]
#endif
static void usage(const char* program, const int exitCode)
{
   fprintf(stderr, "%s %s time_string [...]"
           " [-T|--template time_format_template]"
           " [-F|--float]"
           " [-I|--integer]"
           " [-H|--human-readable]"
           " [-s|--seconds]"
           " [-m|--milliseconds]"
           " [-u|--microseconds]"
           " [-n|--nanoseconds]"
           " [-h|--help]"
           " [-v|--version]\n",
           gettext("Usage:"), program);
   exit(exitCode);
}



// ###### Main program ######################################################
int main(int argc, char** argv)
{
   // ====== Initialise i18n support ========================================
   if(setlocale(LC_ALL, "") == nullptr) {
      setlocale(LC_ALL, "C.UTF-8");   // "C" should exist on all systems!
   }
   setlocale(LC_NUMERIC, "C");   // Use "." for fractional numbers!
   bindtextdomain("time2unixts", SYSTEMTOOLS_LOCALEDIR);
   textdomain("time2unixts");

   // ====== Handle arguments ===============================================
   static const struct option long_options[] = {
      { "template",               required_argument, 0, 'T' },
      { "float",                  no_argument,       0, 'F' },
      { "integer-decimal",        no_argument,       0, 'I' },
      { "integer-hexadecimal",    no_argument,       0, 'X' },
      { "integer-0x-hexadecimal", no_argument,       0, '0' },
      { "human-readable",         no_argument,       0, 'H' },
      { "seconds",                no_argument,       0, 's' },
      { "milliseconds",           no_argument,       0, 'm' },
      { "microseconds",           no_argument,       0, 'u' },
      { "nanoseconds",            no_argument,       0, 'n' },
      { "help",                   no_argument,       0, 'h' },
      { "version",                no_argument,       0, 'v' },
      {  nullptr,                 0,                 0, 0   }
   };

   int          option;
   int          longIndex;
   int          useInteger         = 10;
   bool         humanReadable      = false;
   unsigned int divideBy           = 1;
   const char*  unit               = "ns";
   const char*  timeFormatTemplate = "%Y-%m-%d %H:%M:%S";
   while( (option = getopt_long(argc, argv, "T:FIX0Hsmunvh", long_options, &longIndex)) != -1 ) {
      switch(option) {
         case 'T':
            timeFormatTemplate = optarg;
          break;
         case 'F':
            useInteger = 0;
          break;
         case 'I':
            useInteger = 10;
          break;
         case 'X':
            useInteger = 16;
          break;
         case '0':
            useInteger = -16;
          break;
         case 'H':
            humanReadable = true;
          break;
         case 'n':
            divideBy = 1;
            unit     = "ns";
          break;
         case 'u':
            divideBy = 1000;
            unit     = "µs";
          break;
         case 'm':
            divideBy = 1000000;
            unit     = "ms";
          break;
         case 's':
            divideBy = 1000000000;
            unit     = "s";
          break;
         case 'v':
            version();
          break;
         case 'h':
         case '?':
            // Exit with 0 on h/help, exit with 1 on '?' (unknown option):
            usage(argv[0], (option == 'h') ? 0 : 1);
          break;
         case '-':
          break;
         default:
            // This should not happen: wrong getopt parameters, or missing case?
            fprintf(stderr, "INTERNAL ERROR: Unhandled option c=%c code=%x!\n",
                    (isprint(option) ? (char)option : ' '), option);
            return 1;
          break;
      }
   }

   // ====== Obtain Unix timestamp in ns ====================================
   struct timespec ts;
   for(int i = optind; i <= argc; i++) {
      // ====== Use current time, if no date/time string is given ===========
      if(i == argc) {
         if(optind == argc) {
            if(clock_gettime(CLOCK_REALTIME, &ts) == -1) {
               perror(gettext("clock_gettime() failed"));
               exit(1);
            }
         }
         else {
            break;
         }
      }

      // ====== Parse the next date/time string =============================
      else {
         struct tm   t                  = { 0 };
         const char* timeString         = argv[i];
         const char* remainder          = nullptr;
         const char* secondsPlaceholder = findSecondsPlaceholder(timeFormatTemplate);
         if(secondsPlaceholder != nullptr) {
            long long nanoseconds = 0;

            // ------ Parse front part (before seconds) ---------------------
            const size_t frontLength = (size_t)(secondsPlaceholder - timeFormatTemplate);
            char* frontFormatString  = malloc(frontLength + 1);
            if(frontFormatString == nullptr) {
               fprintf(stderr, gettext("ERROR: malloc() failed: %s!"), strerror(errno));
               exit(1);
            }
            memcpy(frontFormatString, timeFormatTemplate, frontLength);
            frontFormatString[frontLength] = 0x00;
            const char* remainder1 = strptime(timeString, frontFormatString, &t);
            free(frontFormatString);

            // ------ Parse the middle part (seconds) -----------------------
            if(remainder1 != nullptr) {
               char*  remainder2 = nullptr;
               double seconds    = strtod(remainder1, &remainder2);
               if( (remainder2 != remainder1) && (seconds >= 0.0) ) {
                  // Split seconds into integer seconds and nanoseconds:
                  t.tm_sec = (int)seconds;
                  const double fraction = seconds - t.tm_sec;
                  nanoseconds = (long long)(fraction * 1e9 + 0.5);   // +0.5 for rounding safety
                  if(nanoseconds >= 1000000000LL) {
                     t.tm_sec += 1;
                     nanoseconds -= 1000000000LL;
                  }

                  // ------ Parse the back part (after seconds) -------------
                  const char* backFormatString = secondsPlaceholder + 2;
                  if(backFormatString[0] != 0x00) {
                     remainder = strptime(remainder2, backFormatString, &t);
                  } else {
                     remainder = remainder2;
                  }
               }
            }
            ts.tv_sec  = timegm(&t);
            ts.tv_nsec = nanoseconds;
         }
         else {
            remainder  = strptime(argv[i], timeFormatTemplate, &t);
            ts.tv_sec  = timegm(&t);
            ts.tv_nsec = 0;
         }

         if( (remainder == nullptr) || (remainder[0] != 0x00) ) {
            fprintf(stderr, gettext("ERROR: Unable to parse time string \"%s\" with format template \"%s\"!"),
                    timeString, timeFormatTemplate);
            fputs("\n", stderr);
            exit(1);
         }
      }

      // ====== Convert timespec to Unix timestamp ==========================
#if (defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 202311L) && defined(__BITINT_MAXWIDTH__) && (__BITINT_MAXWIDTH__ >= 128))
      const _BitInt(128) unixTS = ((_BitInt(128))1000000000LL * ts.tv_sec) + ts.tv_nsec;
#else
      // NOTE: A 64-bit signed long long will overflow on April 11, 2262!
      const long long unixTS = (1000000000LL * ts.tv_sec) + ts.tv_nsec;
#endif
      if(useInteger == 16) {
         printBigInteger(unixTS / divideBy, 16, false);
      }
      else if(useInteger == -16) {
         printBigInteger(unixTS / divideBy, 16, true);
      }
      else {
         printBigInteger(unixTS / divideBy, 10, false);
         if(useInteger == 0) {
            const unsigned int fractionalUnixTS =
               (double)(unixTS - (unixTS / divideBy) * divideBy);
            const char* format;
            if(divideBy == 1000000000) {
               format = ".%09u";
            }
            else if(divideBy == 1000000) {
               format = ".%06u";
            }
            else if(divideBy == 1000) {
               format = ".%03u";
            }
            else {
               format = ".%u";
            }
            printf(format, fractionalUnixTS);
         }
      }
      if(humanReadable) {
         printf(gettext(" %s since the Unix Epoch"), unit);
      }
      fputs("\n", stdout);
   }

   return 0;
}
